### DISCUSSION TITLE
Piecewise constant material assignment given elemental centroid

### DISCUSSION THREAD

**Discussion by aaelmeli** [idaholab/moose#19094] [Q&A General] [answered]
Hi

I am trying to define the spatial distribution of material given the centroid of each element. The way I am thinking of is that, given the coordinate of the centroid of each element (x,y,z), and the value of the material at this point, each QP in this element will copy that value. 
I have tried to use the `PiecewiseMulticonstant` to achieve this, but obviously, it does some interpolation that I do not want.
For example, in the following picture, this is the output of a simple domain with 2-elements (in 2D). I wanted to assign a value of 1 for the first element, and 2 for the second element. But the results show that element one has a value of 1 and element 2 has a value of 1.5. 
![image](https://user-images.githubusercontent.com/88537645/137173070-1fde53c6-1f66-4b9e-926b-bcc9bcb2da97.png)

The part of the input file associated with this output is as below:

> [Mesh]
>   type = GeneratedMesh
>   dim = 2
>   xmin = 0
>   xmax = 1
>   ymin = 0
>   ymax = 1.0
>   nx = 2
>   ny = 1
> []

> [material_distribution_function]
>     type = PiecewiseMulticonstant
>     data_file = material_distribution.data
>     direction = 'left right'
> []

where the `data_file` is as follows:

>  AXIS X
> 0.25 0.75
> 
> AXIS Y
> 0.5
> 
> DATA
> 1.00
> 2.00

Any hint regarding this?

**Comment by cpgr** [comment, marked as answer]
Hmm, I didn't think `PiecewiseMulticonstant` does any interpolation at all, so this is weird.

You are telling the function to read from the left for the x values, so if you replace your `AXIS X` data with `0 0.5` it should work as you expect.

**Reply by aaelmeli** [reply]
Thank you this fixes the problem. However, I saw an example that defines the material at the centroid of each element [here](https://mooseframework.inl.gov/modules/porous_flow/heterogeneous_models.html) although it uses the `PiecewiseMultilinear`, so that was the confusion, I guess.

But, I am still not sure why there is no function that sets the values based on the centroid of the elements though. In that case, specifying the `direction` for each axis will no longer be needed. And I believe it will be useful for non-structured mesh as well.

@cpgr @GiudGiud Do you think we can work around this and add an option to make `PiecewiseMulticonstant` work without the need to specify the `direction` (given the centroid)

**Reply by GiudGiud** [reply]
I dont see how to get rid of the direction. For elemental fields, the centroid of each element is compared to coordinates specified in data_file. We have to decide whether we get the value on the right or on the left.

**Reply by cpgr** [reply]
I wrote that example that you linked to. Looking at it again, it might be useful to replace `PiecewiseMultilinear` with `PiecewiseMulticonstant` in the example as it is really setting a constant property in each material.

What you are suggesting with a function that sets a value based on element centroid seems like a specialisation of `PiecewiseMulticonstant` that would be better as a separate function.

**Reply by GiudGiud** [reply]
oh are you talking about using some sort of nearest point/neighbor scheme to determine the value?

**Reply by aaelmeli** [reply]
> oh are you talking about using some sort of nearest point/neighbor scheme to determine the value?

Yes, so for example, a user provides the centroid of the element and the value. Then, within each element, instead of comparing the centroid to the coordinates of the mesh as you mentioned, we can compare it to the centroids provided by users. So, in this case, no need to specify the directions. This is just a suggestion, I am not aware of other moose's concerns that may arise /be there.

**Reply by GiudGiud** [reply]
Then you would provide a value for every element? Or interpolate in between.

**Reply by aaelmeli** [reply]
> Then you would provide a value for every element? Or interpolate in between.

For the `PiecewiseMulticonstant `I would provide value for each element if it is one-to-one mapping, Yes. And, if it is not one-to-one, or when mesh refinement is needed, we can always compare the centroids of the finite elements with the centroids provided by the users as well.

**Reply by GiudGiud** [reply]
Ok i ll look into this. 
I think someone had started implementing this, with values coming from CSV as well

**Reply by GiudGiud** [reply]
ok so ElementPropertyFileReader already does all we would want, but:
- it's in tensor mechanics
- it's a UO rather than a function
I ll make a PR soon to move it to the framework & add a function

**Reply by aaelmeli** [reply]
> ok so ElementPropertyFileReader already does all we would want, but:
> 
> * it's in tensor mechanics
> * it's a UO rather than a function
>   I ll make a PR soon to move it to the framework & add a function

That would be great, thank you.

**Reply by Traiwit** [reply]
> ok so ElementPropertyFileReader already does all we would want, but:
> 
> * it's in tensor mechanics
> * it's a UO rather than a function
>   I ll make a PR soon to move it to the framework & add a function


Hi @GiudGiud 
it will be nice if you can add a nodal base 'PropertyFileReader' too, right now it only can assign properties to element, grain and block.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
@aaelmeli @Traiwit could you please take a look at PiecewiseConstantFromCSV in this branch and see if this works for your case. 
Having you test it would be great before we merge it in the framework. That way we merge something that is useful and not just "how far I got implementing something"

@Traiwit  I added nodes to the FileReader
This branch: https://github.com/idaholab/moose/pull/19118
This example to use it: test/tests/functions/piecewise_constant_from_csv/piecewise_constant.i
https://github.com/idaholab/moose/blob/fd37a59fcca6a69b355dd716d60022a23832b2a3/test/tests/functions/piecewise_constant_from_csv/piecewise_constant.i

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Thanks! I will give it a go when I'm at the office tmr.

but I had a quick look and have a quick question:

Since now it is moved to function (from UO), how do I call it from my source code (.C)?

below is how I take the data from .csv to [Materials]

```
params.addParam<UserObjectName>("read_prop_user_object",
                                      "The ElementReadPropertyFile "
                                      "GeneralUserObject to read element "
                                      "specific property values from file");


_read_prop_user_object(isParamValid("read_prop_user_object")
                             ? &getUserObject<ElementPropertyReadFile>("read_prop_user_object")
                             : nullptr),


  _mat_prop[_qp](0) = _read_prop_user_object->getData(_current_elem, 0);
  _mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
  _mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);
```

Thank again @GiudGiud, this would defs be really useful in our workflow.

**Reply by GiudGiud** [reply]
so the UO is still there. The function calls the UO.
In your source code, just call it PropertyReadFile instead of ElementPropertyReadFile. Then call getNodeData instead of getData since you mentioned your data was stored by nodes rather than by element.

**Reply by Traiwit** [reply]
awesome, thanks!

**Reply by aaelmeli** [reply]
> @aaelmeli @Traiwit could you please take a look at PiecewiseConstantFromCSV in this branch and see if this works for your case. Having you test it would be great before we merge it in the framework. That way we merge something that is useful and not just "how far I got implementing something"
> 
> @Traiwit I added nodes to the FileReader This branch: #19118 This example to use it: test/tests/functions/piecewise_constant_from_csv/piecewise_constant.i https://github.com/idaholab/moose/blob/fd37a59fcca6a69b355dd716d60022a23832b2a3/test/tests/functions/piecewise_constant_from_csv/piecewise_constant.i

Thank you @GiudGiud for considering these important changes.
I have tried to test the `PiecewiseConstantFromCSV` in the `nearest` mode. I have exact same mesh and material distribution as the original example in this forum. However, I do not know how to output the material distribution as `auxvariable`. The way I was doing it in case using the `PiecewiseMulticonstant ` is through `FunctionAux` defined in `AuxKernel` and operates on an `auxvariable` . So, I have tried to do the same thing when using `PiecewiseConstantFromCSV`  e.g.

>    [material_FromCSV_auxkernel]
>     type = FunctionAux
>     function = nearest
>     variable = material_FromCSV_auxvariable
>   []

but I am getting an error telling me that the "material_FromCSV_auxvariable" does not exist in the system. Any thoughts on how to achieve this to continue testing and visually check the results?

**Reply by GiudGiud** [reply]
What do you have in your [AuxVariables] block? It seems `material_FromCSV_auxvariable` has not been defined. 

FunctionAux + PiecewiseConstantFromCSV + PropertyReadFile UserObject should be the way to go. 
You can also use a FunctionIC instead of a FunctionAux since I m guess that aux variable is not changing during the simulation

**Reply by aaelmeli** [reply]
> What do you have in your [AuxVariables] block? It seems `material_FromCSV_auxvariable` has not been defined.
> 
> FunctionAux + PiecewiseConstantFromCSV + PropertyReadFile UserObject should be the way to go. You can also use a FunctionIC instead of a FunctionAux since I m guess that aux variable is not changing during the simulation

I found the mistake, yes it is in my `auxvariable `declaration. I was declaring the `material_FromCSV_auxvariable ` as `CONSTANT` and `SCALAR`. I changed it to `MONOMIAL ` and it works as it should. Thank you.

**Reply by Traiwit** [reply]
@GiudGiud did you merge this function with this month's MOOSE update? it would be very helpful to others too I reckon :)

**Reply by GiudGiud** [reply]
not yet. It did not get reviewed yet
